rs-matter/src/data_model/cluster_window_covering.rs

```diff
@@ -0,0 +1,658 @@
+        CommandsDiscriminants::StopMotion as _,
+        CommandsDiscriminants::GoToLiftValue as _,
+        CommandsDiscriminants::GoToLiftPercentage as _,
+        CommandsDiscriminants::GoToTiltValue as _,
+        CommandsDiscriminants::GoToTiltPercentage as _,
+    ],
+};
+
+pub struct WindowCoveringCluster {
+    data_ver: Dataver,
+    type_attribute: Cell<TypeAttribute>,
+    config_status: Cell<ConfigStatus>,
+    operational_status: Cell<OperationalStatus>,
+    end_product_type: Cell<EndProductType>,
+    mode: Cell<Mode>,
+}
+
+impl WindowCoveringCluster {
+    pub fn new(rand: Rand) -> Self {
+        Self {
+            data_ver: Dataver::new(rand),
+            type_attribute: Cell::new(TypeAttribute::default()),
+            config_status: Cell::new(ConfigStatus::default()),
+            operational_status: Cell::new(OperationalStatus::default()),
+            end_product_type: Cell::new(EndProductType::default()),
+            mode: Cell::new(Mode::default()),
+        }
+    }
+
+    pub fn set_type(&self, device_type: TypeAttribute) {
+        if self.type_attribute.get() != device_type {
+            self.type_attribute.set(device_type);
+            self.data_ver.changed();
+        }
+    }
+
+    pub fn read(&self, attr: &AttrDetails, encoder: AttrDataEncoder) -> Result<(), Error> {
+        if let Some(writer) = encoder.with_dataver(self.data_ver.get())? {
+            if attr.is_system() {
+                CLUSTER.read(attr.attr_id, writer)
+            } else {
+                match attr.attr_id.try_into()? {
+                    Attributes::Type(codec) => codec.encode(writer, self.type_attribute.get()),
+                    Attributes::ConfigStatus(codec) => {
+                        codec.encode(writer, self.config_status.get())
+                    }
+                    Attributes::OperationalStatus(codec) => {
+                        codec.encode(writer, self.operational_status.get())
+                    }
+                    Attributes::EndProductType(codec) => {
+                        codec.encode(writer, self.end_product_type.get())
+                    }
+                    Attributes::Mode(codec) => codec.encode(writer, self.mode.get()),
+                    _ => todo!("read {:?}", attr.attr_id),
+                }
+            }
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn write(&self, attr: &AttrDetails, data: AttrData) -> Result<(), Error> {
+        let data = data.with_dataver(self.data_ver.get())?;
+
+        match attr.attr_id.try_into()? {
+            Attributes::Type(codec) => self.set_type(codec.decode(data)?),
+            _ => todo!("write {:?}", attr.attr_id),
+        }
+
+        self.data_ver.changed();
+
+        Ok(())
+    }
+
+    pub fn invoke(
+        &self,
+        _exchange: &Exchange,
+        cmd: &CmdDetails,
+        _data: &TLVElement,
+        _encoder: CmdDataEncoder,
+    ) -> Result<(), Error> {
+        match cmd.cmd_id.try_into()? {
```

Comment:
Ditto. Would that crash?

Reply:
Ditto. Errors should be propagated up the call stack. Do you disagree how it is handled here?

Reply:
`todo!` panics. It does not propagate errors up the call stack. It will essentially crash your app.